# Cattle-Genome-Functional-Annotation
We established the first global map of regulatory elements (15 chromatin states) and defined their coordinated activities in cattle, through genome-wide profiling for six histone modifications, RNA polymerase II, CTCF-binding sites, DNA accessibility, DNA methylation, and transcriptome in Rumen Epithelial Primary Cells (REPC), rumen tissues, and Madin-Darby Bovine Kidney Epithelial Cells (MDBK). We demonstrated that each chromatin state exhibited specific enrichments for sequence ontology, gene expression, methylation, trait-associated variants, eQTLs, selection signatures, and evolutionary conserved elements, implying distinct biological functions. Our results show the crucial role of functional genome annotation for understanding gene regulation, complex traits variation, domestication, and adaptive evolution in livestock.    

All the peaks of studied epigenomic marks across 38 samples can be accessed via the All_peaks.gz file, and the predicted chromatin states can be downloaded (ChromHMM_REPC.bed) here as well.
